### PR TITLE
Support HIGHESTMODSEQ and VANISHED

### PR DESCRIPTION
**Pull request by mordak**
This adds support for imap-proto `ResponseCode::HighestModSeq(u64)` and the `VANISHED` response described in RFC 7162 (QRESYNC). With these extensions, it is possible to synchronize a mailbox in a single round trip using `UID FETCH` and passing a modifier `(CHANGEDSINCE <MODSEQ> VANISHED)`. Only modifications and new messages since `<MODSEQ>` will be returned, and deleted messages will be returned in a `VANISHED` response on the `unsolicited_responses` channel.

Technically, if `QRESYNC` is enabled then `VANISHED` responses will be sent to `EXPUNGE` commands. Presently these will show up on the `unsolicited_responses` channel even though they are directly a response to a command. I wasn't sure what the best way to handle this would be in `parse_expunge()`, and commented to this effect in `parse_vanished_test()`, and the test is constructed in a way to verify that `parse_expunge()` succeeds under this condition, and the `VANISHED` response is available in the `unsolicited_responses` channel.

I also fixed clippy warnings. One of them I actually introduced (useless format string) - sorry. The other one is for the [manual_non_exhaustive lint](https://rust-lang.github.io/rust-clippy/master/index.html#manual_non_exhaustive). I separated them out into different commits so you can see them in isolation.

I am happy to revise this if you would like, particularly if you have a preference for what to do in `parse_expunge()`. I did experiment with expanding the `VANISHED` uid sequence-set out into a `Vec<u32>`, but it's kind of ugly, and also kind of dangerous for very large ranges, so I erred on the side of at least being consistent and having these responses always show up on the `unsolicited_responses` channel.

---
This change is [Reviewable](https://reviewable.io/reviews/jonhoo/rust-imap/172)

### CODE REVIEW

**Comment by jonhoo** [comment]
Thank you for this, those are good changes!

Sadly, these changes are also backwards-incompatible, and would require a major version bump (see #163 and https://github.com/jonhoo/rust-imap/pull/149#discussion_r426815938), so I may have to hold off on merging it for a little while. I'll leave some other comments in the meantime though!

As for `parse_expunge`, I'm not sure what the right way to go about it is either. I _think_ having the `VANISHED` responses go through unilateral responses is probably fine. It's not ideal as far as the API goes, but at least it makes it _possible_ for clients to do what they need to.

File: src/types/mailbox.rs
Comment (jonhoo) [inline src/types/mailbox.rs:41]: We should probably add `#[non_exhaustive]` to `Mailbox` as well so that we can make these kinds of changes in the future without a breaking change.

File: src/types/mailbox.rs
Comment (jonhoo) [inline src/types/mailbox.rs:39]:
```suggestion
    /// The highest mod sequence for this mailbox. Used with
```

File: src/types/mailbox.rs
Comment (mordak) [reply src/types/mailbox.rs:42]: Ah, that is a good idea! I added this change.

**Comment by mordak** [comment]
> Sadly, these changes are also backwards-incompatible, and would require a major version bump (see #163 and [#149 (comment)](https://github.com/jonhoo/rust-imap/pull/149#discussion_r426815938)), so I may have to hold off on merging it for a little while. I'll leave some other comments in the meantime though!

Yup, no problem. If you are looking for an excuse to take the plunge on the major bump, then this is the only outstanding diff before I can publish [IMAP <-> Maildir sync tool](https://github.com/mordak/runt). It has been a private project for a long time though, so I am not in any rush. :-)

> As for `parse_expunge`, I'm not sure what the right way to go about it is either. I _think_ having the `VANISHED` responses go through unilateral responses is probably fine. It's not ideal as far as the API goes, but at least it makes it _possible_ for clients to do what they need to.

Yeah. If `parse_expunge` keeps the current signature, then it's expanding the `VANISHED` UID sequence-set out into a `Vec<u32>`. Otherwise, the signature changes to return an enum type with a `Vec<u32>` or `Vec<RangeInclusive>`; or the signature changes to return something else. I think the enum type is the most correct way to do it (they _are_ different things after all, `id` vs `uid` values, etc.), but it doesn't quite feel right? Of course, if you have a preference, then getting it in at the same time you're bumping the major anyway would be nice.

**Comment by mordak** [comment]
> As for `parse_expunge`, I'm not sure what the right way to go about it is either. I _think_ having the `VANISHED` responses go through unilateral responses is probably fine. It's not ideal as far as the API goes.

Thinking about this some more, I think the nicest API for the client is for `client::expunge()` and `client::uid_expunge()` to return something that they can iterate easily. So if those returned (say) `Result<Expunged>` and `Expunged` was an enum (or something) that provided an iterator interface, then the client could do something nice like `for ids in client.expunge()? { ... }`. If the iterator just yielded `u32`s, then some clients might not even notice, since you can do the exact same thing with a `Vec`.

I am happy to work it out if you have a preference. Since I'm the one who wants `VANISHED` support, I should do the legwork to ensure it fits in the API nicely for you.

**Comment by jonhoo** [comment]
> Yup, no problem. If you are looking for an excuse to take the plunge on the major bump, then this is the only outstanding diff before I can publish my [IMAP <-> Maildir sync tool](https://github.com/mordak/runt). It has been a private project for a long time though, so I am not in any rush. :-)

That's really cool! That definitely does give me a bit of a push :p

I like your idea of returning just "something iterable" instead — that does seem a lot nicer. Alternatively it could return a `Vec<enum>` where the `enum` is either a single id or a sequence-set?

**Comment by mordak** [comment]
> I like your idea of returning just "something iterable" instead — that does seem a lot nicer. Alternatively it could return a `Vec<enum>` where the `enum` is either a single id or a sequence-set?

Yup, `Vec<enum>` would also work. I think the caller would have to `match` on the enum though to get at the values? The upside of this way is the caller will know if they're getting `ID` or `UID` values, because they would explicitly know if they're dealing with an `EXPUNGE` response (ids) or a `VANISHED` response (uids). I think there is value in making this distinction to the caller - `UID` and `ID` have different semantics (all the business around IDs decrementing between responses like in the example [in RFC4315](https://tools.ietf.org/html/rfc4315#section-2.1) is stuff the caller has to deal with).

I will mock something up and shoot you a diff. If you like it, I can merge it into this PR. If you don't, then we can hammer it out. :-)

**Comment by mordak** [comment]
> I like your idea of returning just "something iterable" instead — that does seem a lot nicer. Alternatively it could return a `Vec<enum>` where the `enum` is either a single id or a sequence-set?

Pushed a commit with an implementation of a `Deleted` type that is returned from `parse_expunge()`.

The client can just match on the enum if they want to get at the underlying `Vec<Seq>` or `Vec<RangeInclusive<Uid>>`, but I also provided a few iterator conveniences and some examples for how they can be used. There's 3 ways to iterate:
- `Deleted::into_iter()`. So `for id in &deleted { } ` works and the client doesn't care what the numbers are (sequence numbers or UIDs). I figure this is probably compatible with what most users are probably already doing?
- `Deleted::seqs()`, if the client knows / expects to get an `EXPUNGE` response.
- `Deleted::uids()`, if the client knows / expects to get a `VANISHED` response.

Let me know what you think. I am happy to modify / rework if needed.

File: src/parse.rs
Comment (jonhoo) [inline src/parse.rs:603]: Should this also assert that `seqs()` is empty?

File: src/types/deleted.rs
Comment (jonhoo) [inline src/types/deleted.rs:66]: I think you could probably make this and `uids` much nicer by returning `impl ExactSizeIterator<Item = u32>`. That way you can get rid of all the various `struct` types and `impl Iterator for` blocks below, and just use combinators instead (`flat_map` will probably come in handy for `uids` for example).

I also don't think the `Option` adds much here. `ExactSizeIterator` is good enough since it also provides an `is_empty` method.

**Comment by jonhoo** [comment]
This looks great — I think having the one shared type is a great improvement. I left a note on how you may be able to (significantly) simplify the implementation though!

File: src/parse.rs
Comment (mordak) [reply src/parse.rs:606]: Ah, yes!

File: src/types/deleted.rs
Comment (mordak) [reply src/types/deleted.rs]: Oh, this is much better - thanks! I had not thought to do it this way.

I pushed a new version with `seqs()` and `uids()` returning `impl Iterator<Item = u32>`. `ExactSizeIterator` isn't implemented for `FlatMap` (which makes sense, I suppose), so I used `Iterator` instead. I could use `ExactSizeIterator` for `seqs()`, but wanted to keep them consistent.

As for the `Option` - I included it to let the caller distinguish the 'no values' from the 'didn't get that kind of response' cases, but wasn't sure if this was a distinction that added enough value to justify the `Option`, so am glad you have an opinion! I removed it in this version, though `Iterator` doesn't provide `is_empty()` like `ExactSizeIterator` does, but I am not entirely convinced this is needed? `Deleted` has `is_empty()` anyway. If you think this is important I can figure it out.

Thanks for the comments - I found all the structs a bit cumbersome, but didn't think to just return an `impl`. Being able to delete all that makes it much simpler / nicer.

Comment (mordak) [reply src/types/deleted.rs]: Also, TIL `flat_map()` is a thing. That is much nicer than rolling it by hand!

File: src/parse.rs
Comment (jonhoo) [inline src/parse.rs:161]: Can there be both vanished and expunged results at once?

File: src/types/deleted.rs
Comment (jonhoo) [inline src/types/deleted.rs:64]: This should probably use the `Seq` type alias to signify that it specifically returns sequence numbers.

File: src/types/deleted.rs
Comment (jonhoo) [inline src/types/deleted.rs:75]: This should probably use the `Uid` type alias to signify that it specifically returns UIDs.

File: src/types/deleted.rs
Comment (jonhoo) [inline src/types/deleted.rs:80]: It doesn't seem quite right for this to need to clone the range. I think `iter().copied()` should do the trick?

**Comment by jonhoo** [comment]
Ah, yes, this is looking much nicer, thank you! I left another couple of notes from looking through more carefully.

I'm not quite sure why the Minimum supported Rust version test is failing. Can you try bumping that to 1.41 here
https://github.com/jonhoo/rust-imap/blob/7868e312ff1349a667541b235053fc5c95c94854/azure-pipelines.yml#L50
and here
https://github.com/jonhoo/rust-imap/blob/7868e312ff1349a667541b235053fc5c95c94854/azure-pipelines.yml#L55

Also, while you're at it, could you please update the changelog to reflect both the Rust version change and the changes in this PR?

File: src/parse.rs
Comment (mordak) [reply src/parse.rs:161]: According to the spec, no. Once the client has enabled QRESYNC then the server must send only VANISHED responses in lieu of EXPUNGE responses.

https://tools.ietf.org/html/rfc7162#section-3.2.10

File: src/types/deleted.rs
Comment (mordak) [reply src/types/deleted.rs]: You're right, fixed!

File: src/types/deleted.rs
Comment (mordak) [reply src/types/deleted.rs]: Fixed!

File: src/types/deleted.rs
Comment (mordak) [reply src/types/deleted.rs:80]: I actually fiddled with this a bunch and can't get it to work without the `clone()`. `RangeInclusve` is not `Copy`, and doesn't implement `iter()` because it _is_ the iterator. We can't just iterate the one in the `Vec` because it means mutating the `Vec` members in place (ie. [`next()` mutates the `RangeInclusive`](https://doc.rust-lang.org/src/core/iter/range.rs.html#649)).  So inside `flat_map()` we're expected to yield an iterator from the lambda, but because we cannot mutate the `RangeInclusive` in the `Vec` directly, we need a new iterator to return inside `flat_map()`, which means making a new one from the given `RangeInclusive`, which is a `clone()`. 

But I think this is really not a big deal. `RangeInclusive<u32>` is tiny (a pair of numbers and some state), so cloning it is cheap, and it means we do not mutate the members of the `Delete`, so we can iterate it more than once if we want, and I think that's the right behaviour to have.

**Comment by mordak** [comment]
> Ah, yes, this is looking much nicer, thank you! I left another couple of notes from looking through more carefully.

Thanks! Your comments are really helpful.

> I'm not quite sure why the Minimum supported Rust version test is failing. Can you try bumping that to 1.41 here

Figured this out. It is because nom6 brings in a new version of bitvec, which has a minimum version of 1.44. I tested it with 1.43 and it works, so I set that as the minimum version. If you want to go to 1.44 then that's easy - it works with every version between 1.43 and the latest stable.

We actually do not _need_ the bump to nom6. I only did it because imap-proto did it, and it makes it possible to work on a local imap-proto while we're adding features to it and plumbing them through this crate. So when imap-proto went to nom6, I bumped this crate to nom6 so I could work on them in tandem (via `[patch.crates-io]`). I can revert it to nom5 again if you would prefer and I think this would let the minimum version stay at 1.40, but I am not sure if this is a priority for you. I totally get if it is, so just say so and I can figure it out.

> Also, while you're at it, could you please update the changelog to reflect both the Rust version change and the changes in this PR?

Done! Let me know if I messed it up or was too brief / verbose / anything. :-)

File: CHANGELOG.md
Comment (jonhoo) [inline CHANGELOG.md:12]: Let's say what it increased to :)

File: azure-pipelines.yml
Comment (jonhoo) [inline azure-pipelines.yml:55]: Could you replace the comment here too now that the reason for the MSRV is different.

**Comment by jonhoo** [comment]
I'm very happy to update to nom6, so please do keep that change!

The changes to the changelog look good -- I only left some minor notes and then I think this PR is basically ready :tada: Now I just need to get ready for a new major release :sweat_smile:

File: CHANGELOG.md
Comment (mordak) [reply CHANGELOG.md]: Done!

File: azure-pipelines.yml
Comment (mordak) [reply azure-pipelines.yml]: No problem! :-)

**Comment by mordak** [comment]
> The changes to the changelog look good -- I only left some minor notes and then I think this PR is basically ready 🎉 Now I just need to get ready for a new major release 😅

Great! If there is anything I can help with then let me know, or if you need additional changes here then just say the word.

Thanks for your time and effort helping me get this into shape! :-)

File: src/parse.rs
Comment (jonhoo) [reply src/parse.rs:161]: Ah, excellent! Could you please add a comment (with that link) pointing out that fact?

**Comment by jonhoo** [comment]
Oh, just a note for the future: in general it's best to avoid force pushing during reviews, since GitHub has a hard time keeping track of what changed when you do that. It tends to be easier to do commit cleanup just at the very end instead :)

I missed one last thing that I left as a comment now, and then I think I'm done. Sorry to keep adding more nits — thank you for sticking with me through that!

File: src/parse.rs
Comment (mordak) [reply src/parse.rs:165]: Done!

**Comment by mordak** [comment]
> Oh, just a note for the future: in general it's best to avoid force pushing during reviews, since GitHub has a hard time keeping track of what changed when you do that. It tends to be easier to do commit cleanup just at the very end instead :)

Oh, sorry. :( 

Thank you for saying this! I didn't know this  made reviews harder, and am used to force pushing to keep the commit history tidy. I didn't force the last one here. :-)

Incidentally, if you want some of the commits merged / shuffled / anything, just let me know and I can wrangle it.

> I missed one last thing that I left as a comment now ([#172 (comment)](https://github.com/jonhoo/rust-imap/pull/172#discussion_r537118865)), and then I think I'm done. Sorry to keep adding more nits — thank you for sticking with me through that!

No problem at all! It's your crate and your standards, and you're the one with the merge button. Don't feel bad for nits - if I don't fix them now you're just going to feel compelled to fix them later. :-)

**Review by jonhoo — APPROVED** [review]

**Pull request merged**
